src/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -451,39 +463,64 @@ private int InternalValueCountCore()
 
         /// <summary>Retrieves an array of strings containing all the value names.</summary>
         /// <returns>All value names.</returns>
-        private unsafe string[] GetValueNamesCore(int values)
+        private unsafe string[] GetValueNamesCore()
         {
-            string[] names = new string[values];
-            char[] name = new char[MaxValueLength + 1];
-            int namelen;
+            List<string> names = new List<string>();
 
-            fixed (char* namePtr = &name[0])
+            // Names in the registry aren't usually very long, although they can go to as large
+            // as 16383 characters (MaxValueLength).
+            //
+            // Every call to RegEnumValue will allocate another buffer to get the data from
+            // NtEnumerateValueKey before copying it back out to our passed in buffer. This can
+            // add up quickly- we'll try to keep the memory pressure low and grow the buffer
+            // only if needed.
+
+            char[] name = ArrayPool<char>.Shared.Rent(100);
+
+            try
             {
-                for (int i = 0; i < values; i++)
+                int index = 0;
+                int result;
+                int nameLength = name.Length;
+
+                while ((result = Interop.Advapi32.RegEnumValue(
+                    _hkey,
+                    index,
+                    name,
+                    ref nameLength,
+                    IntPtr.Zero,
+                    null,
+                    null,
+                    null)) != Interop.Errors.ERROR_NO_MORE_ITEMS)
                 {
-                    namelen = name.Length;
-
-                    int ret = Interop.Advapi32.RegEnumValue(_hkey,
-                        i,
-                        namePtr,
-                        ref namelen,
-                        IntPtr.Zero,
-                        null,
-                        null,
-                        null);
-
-                    if (ret != 0)
+                    switch (result)
                     {
-                        // ignore ERROR_MORE_DATA if we're querying HKEY_PERFORMANCE_DATA
```

Comment:
This PERFORMANCE_DATA bit is a mistake. This can happen when retrieving data from that key, not names. In any case, we're handling ERROR_MORE_DATA here.